File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.java
class Solution {
    public String frequencySort(String s) {
       
       int length = s.length();
       int freq[] = new int[256];
       ArrayList<ArrayList<Character>> buckets = new ArrayList<>(length + 1);

       for(int i = 0 ; i <= length ; i++)
            buckets.add(new ArrayList<>());

       for(char ch : s.toCharArray())
           freq[ch]++;

       for(int i = 0 ; i < 256 ; i++)
       {
           int num = freq[i];

           if(num > 0)
                buckets.get(num).add((char)(i));
       }

        StringBuilder result = new StringBuilder();

        for(int i = length ; i >= 0 ; i--)
        {
            if(buckets.get(i) != null)
            {
                for(char ch : buckets.get(i))
                {
                    result.append(String.valueOf(ch).repeat(i));
                }   
            }
        }

        return result.toString();
    }
}

File: 0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.js
/**
 * @param {string} s
 * @return {string}
 */
var frequencySort = function(s) {

    let freq = new Array(256).fill(0);
    let buckets = new Array(s.length + 1);
    let result = "";

    for(let i = 0 ; i <= s.length ; i++)
        buckets[i] = [];

    for(let ch of s)
    {
        freq[ch.charCodeAt(0)]++;   
    }

    for(let i = 0 ; i < 256 ; i++)
    {
        buckets[freq[i]].push(String.fromCharCode(i));
    }

    for(let i = buckets.length - 1 ; i >= 0 ; i--)
    {
        if(buckets[i])
        {
            buckets[i].forEach(el =>
            {
                result += el.repeat(i);
            })
        }
    }

    return result;
};